ctru-rs/src/applets/swkbd.rs

```diff
@@ -90,41 +92,39 @@ impl Swkbd {
     pub fn init(keyboard_type: Kind, num_buttons: i32) -> Self {
         unsafe {
             let mut state = Box::<SwkbdState>::default();
-            swkbdInit(state.as_mut(), keyboard_type as u32, num_buttons, -1);
+            swkbdInit(state.as_mut(), keyboard_type.into(), num_buttons, -1);
             Swkbd { state }
         }
     }
 
     /// Gets input from this keyboard and appends it to the provided string.
     ///
-    /// The text received from the keyboard will be truncated if it is greater than 2048 bytes
-    /// in length.
-    pub fn get_utf8(&mut self, buf: &mut String) -> Result<Button, Error> {
+    /// The text received from the keyboard will be truncated if it is longer than `max_bytes`.
+    pub fn get_string(&mut self, max_bytes: usize) -> Result<(String, Button), Error> {
         // Unfortunately the libctru API doesn't really provide a way to get the exact length
         // of the string that it receieves from the software keyboard. Instead it expects you
         // to pass in a buffer and hope that it's big enough to fit the entire string, so
         // you have to set some upper limit on the potential size of the user's input.
-        const MAX_BYTES: usize = 2048;
-        let mut tmp = [0u8; MAX_BYTES];
-        let button = self.get_bytes(&mut tmp)?;
+        let mut tmp = vec![0u8; max_bytes];
+        let button = self.write_exact(&mut tmp)?;
 
         // libctru does, however, seem to ensure that the buffer will always contain a properly
         // terminated UTF-8 sequence even if the input has to be truncated, so these operations
         // should be safe.
         let len = unsafe { libc::strlen(tmp.as_ptr()) };
-        let utf8 = unsafe { str::from_utf8_unchecked(&tmp[..len]) };
+        tmp.truncate(len);
 
-        // Copy the input into the user's `String`
-        *buf += utf8;
-        Ok(button)
+        let res = unsafe { String::from_utf8_unchecked(tmp) };
+
+        Ok((res, button))
     }
 
     /// Fills the provided buffer with a UTF-8 encoded, NUL-terminated sequence of bytes from
     /// this software keyboard.
     ///
     /// If the buffer is too small to contain the entire sequence received from the keyboard,
-    /// the output will be truncated but should still be well-formed UTF-8
-    pub fn get_bytes(&mut self, buf: &mut [u8]) -> Result<Button, Error> {
+    /// the output will be truncated but should still be well-formed UTF-8.
+    pub fn write_exact(&mut self, buf: &mut [u8]) -> Result<Button, Error> {
```

Comment:
Not sure that `write_exact` is a good name for this, since it's reading not writing

Reply:
I guess there's two ways of looking at it - it's reading from the software keyboard, but writing to the buffer that's passed in. I guess the `std::io::Read` names are based on "reading from self" so perhaps `read_exact` would be more in line with that...

Reply:
Honestly, I believe it should just implement `std::io::Read`. We can document the blocking behaviour in the trait impl.